src/code/z_demo.c

```diff
@@ -391,176 +446,189 @@ void func_80064824(PlayState* play, CutsceneContext* csCtx, CsCmdBase* cmd) {
         }
     }
 }
 
-// Command 9: ?
-void Cutscene_Command_09(PlayState* play, CutsceneContext* csCtx, CsCmdUnknown9* cmd) {
-    if (csCtx->frames == cmd->startFrame) {
-        Rumble_Request(0.0f, cmd->unk_06, cmd->unk_07, cmd->unk_08);
+void CutsceneCmd_RumbleController(PlayState* play, CutsceneContext* csCtx, CsCmdRumble* cmd) {
+    if (csCtx->curFrame == cmd->startFrame) {
+        Rumble_Request(0.0f, cmd->sourceStrength, cmd->duration, cmd->decreaseRate);
     }
 }
 
-// Command 0x8C: Set Time of Day & Environment Time
-void func_80065134(PlayState* play, CutsceneContext* csCtx, CsCmdDayTime* cmd) {
-    s16 temp1;
-    s16 temp2;
+void CutsceneCmd_SetTime(PlayState* play, CutsceneContext* csCtx, CsCmdTime* cmd) {
+    s16 hours;
+    s16 minutes;
 
-    if (csCtx->frames == cmd->startFrame) {
-        temp1 = (cmd->hour * 60.0f) / (360.0f / 0x4000);
-        temp2 = (cmd->minute + 1) / (360.0f / 0x4000);
+    if (csCtx->curFrame == cmd->frame) {
+        hours = (cmd->hour * 60.0f) / (24.0f * 60.0f / 0x10000);
+        minutes = (cmd->minute + 1) / (24.0f * 60.0f / 0x10000);
 
-        gSaveContext.dayTime = temp1 + temp2;
-        gSaveContext.skyboxTime = temp1 + temp2;
+        gSaveContext.dayTime = hours + minutes;
+        gSaveContext.skyboxTime = hours + minutes;
     }
 }
 
-// Command 0x3E8: Code Execution (& Terminates Cutscene?)
-void Cutscene_Command_Terminator(PlayState* play, CutsceneContext* csCtx, CsCmdBase* cmd) {
+void CutsceneCmd_Destination(PlayState* play, CutsceneContext* csCtx, CsCmdDestination* cmd) {
     Player* player = GET_PLAYER(play);
-    s32 temp = 0;
+    s32 titleDemoSkipped = false;
 
     if ((gSaveContext.gameMode != GAMEMODE_NORMAL) && (gSaveContext.gameMode != GAMEMODE_END_CREDITS) &&
-        (play->sceneId != SCENE_HYRULE_FIELD) && (csCtx->frames > 20) &&
+        (play->sceneId != SCENE_HYRULE_FIELD) && (csCtx->curFrame > 20) &&
         (CHECK_BTN_ALL(play->state.input[0].press.button, BTN_A) ||
          CHECK_BTN_ALL(play->state.input[0].press.button, BTN_B) ||
          CHECK_BTN_ALL(play->state.input[0].press.button, BTN_START)) &&
         (gSaveContext.fileNum != 0xFEDC) && (play->transitionTrigger == TRANS_TRIGGER_OFF)) {
         Audio_PlaySfxGeneral(NA_SE_SY_PIECE_OF_HEART, &gSfxDefaultPos, 4, &gSfxDefaultFreqAndVolScale,
                              &gSfxDefaultFreqAndVolScale, &gSfxDefaultReverb);
-        temp = 1;
+        titleDemoSkipped = true;
     }
 
-    if ((csCtx->frames == cmd->startFrame) || (temp != 0) ||
-        ((csCtx->frames > 20) && CHECK_BTN_ALL(play->state.input[0].press.button, BTN_START) &&
+    if ((csCtx->curFrame == cmd->frame) || titleDemoSkipped ||
+        ((csCtx->curFrame > 20) && CHECK_BTN_ALL(play->state.input[0].press.button, BTN_START) &&
          (gSaveContext.fileNum != 0xFEDC))) {
-        csCtx->state = CS_STATE_UNSKIPPABLE_EXEC;
+        csCtx->state = CS_STATE_RUN_UNSKIPPABLE;
```

Comment:
`CS_STATE_RUN_UNSKIPPABLE` being set unconditionally when a destination command is processed, and it being set preventing cutscenes from being stopped, makes me think it's intended as "the cutscene will never end, it will only end as the play state itself ends"

This interpretation seems strengthened by the (only?) destination which doesn't end the play state (by a transition) `CS_DEST_GANONS_CASTLE_DISPEL_BARRIER_CONDITONAL`, setting `CS_STATE_STOP` manually

so "unskippable" -> (something I don't have an idea for atm) ?

`wait_transition` ? idk

Reply:
i think unskippable describes it fine. the destination command does set it, cause it doesnt want you to skip the destination command, but I dont think it needs to be named after it specifically because it is the only use

Reply:
> cause it doesnt want you to skip the destination command

what do you mean?

Reply:
the destination command sets unskippable so that it will guarantee it will be processed. you can mash the debug skip cutscene button all you want, but it will not skip the destination command

Reply:
Which debug skip cutscene button?

Reply:
```
    if (CHECK_BTN_ALL(play->state.input[0].press.button, BTN_DRIGHT)) {
        csCtx->state = CS_STATE_STOP;
        return;
    }
```

Reply:
theres also some logic for START being used as a skip cutscene button in the destination function
their on screen print tells you about both
`GfxPrint_Printf(&printer, "%s", "SKIP=(START) or (Cursole Right)");`

Reply:
https://github.com/zeldaret/oot/pull/1327#discussion_r1048991206

Where is `CS_STATE_RUN_UNSKIPPABLE` relevant for this?



---------

> theres also some logic for START being used as a skip cutscene button in the destination function their on screen print tells you about both `GfxPrint_Printf(&printer, "%s", "SKIP=(START) or (Cursole Right)");`

afaict the start check in `CutsceneCmd_Destination` makes the destination be applied early, effectively skipping the cutscene

Reply:
right, and when the destination command runs, it sets unskippable so other things cant skip it 

my point is that the important property is that it wont be skipped, not that its waiting for a transition. "waiting for the transition" is the single use in the game where unskippable is set, but its not the only thing it could be used for. when describing its purpose, it exists as a version of run that cannot be skipped

Reply:
What?

How can the destination command be "skipped" anyway with already having been executed?

Reply:
Maybe a better question:
What does `CS_STATE_RUN_UNSKIPPABLE` make "unskippable" ?

Reply:
Another maybe better question:
Where are the skips implemented and how does `CS_STATE_RUN_UNSKIPPABLE` prevent them?

Reply:
anything that sets this state. which in the final game is only the destination command as you noted, but it doesnt have to be only for that.
the important point is that it gets checked against before allowing "stop" operations to go through (Cutscene_StopManual/CS_MISC_STOP_CUTSCENE)

Reply:
Are we on the same page that when `CS_STATE_RUN_UNSKIPPABLE` is set, the current play state instance has seconds to live?

Reply:
*in the only case the final game uses unskippable* yes, this is true. 
looking wider picture, the mode exists to prevent early stoppage/skipping of the cutscene

Reply:
I don't understand how you're seeing things

Here's a summary of how I see things:

A cutscene script is active, including a `CS_DESTINATION`
That means `CutsceneCmd_Destination` runs every frame

In `CutsceneCmd_Destination`:

There is a `titleDemoSkipped` temp, `false` by default, which is set to `true` if A/B/Start is pressed in one of the title demo "attract mode" cutscenes.
Other than that the whole function body is under a condition that passes if either of these passes:
- The cutscene is at the `CS_DESTINATION`-specified frame
- `titleDemoSkipped` is true
- The start button is pressed (this is a debug feature, the first so far in this summary)

In the body below that condition, `CS_STATE_RUN_UNSKIPPABLE` is set, and the destination is handled in a giant switch:
all cases except one start a scene transition, meaning the current play state instance will end soon :tm: (in a few seconds)
The one case that doesn't (`CS_DEST_GANONS_CASTLE_DISPEL_BARRIER_CONDITONAL`) sets the state to `CS_STATE_STOP` instead

When the condition under which most of the function body is runs, that basically puts an end to the cutscene.
That happens either
- at the normal end of the cutscene, if the `CS_DESTINATION`-specified frame is reached,
- if the title screen demo "attract mode" cutscene is skipped by pressing A/B/Start (`titleDemoSkipped` is true)
- if start is pressed, as a debug feature to skip cutscenes

Reply:
I completely agree with this analysis of its use in the destination command

our disagreement comes from wanting to name this feature after its only apparent use in the game, vs what its overall effect on the system is

i view this modes existence as a variation of the "run" mode which other things will stop the system from being able to go into the "stop" mode. In the final game, the destination command does do this. But it does not only have to be used in this manner

Reply:
What I observe is `CS_STATE_RUN_UNSKIPPABLE` is set exactly when the cutscene puts an end to itself by the means of a scene transition
So I would name it after that: the cutscene lasting until the end of the play state

I don't understand what supports naming it "unskippable" since I don't see its connection with any skipping

Reply:
the cutscene cant be ended in methods that check against it (which admittedly is not all of them, but they didnt consistently guard skip/stop checks with `CS_STATE_RUN_UNSKIPPABLE` everywhere)

we commonly have this disagreement where you take the very objective approach to naming things based on only its existing usages, where I think its fair to extrapolate on a features overall meaning and name it accordingly